aten/src/ATen/native/mps/operations/FusedAdamWKernelImpl.mm

```diff
@@ -39,5 +39,35 @@ void _fused_adamw_mps_impl_(at::TensorList params,
                                                  eps,
                                                  maximize);
 }
+
+void _fused_adamw_mps_impl_(at::TensorList params,
+                            at::TensorList grads,
+                            at::TensorList exp_avgs,
+                            at::TensorList exp_avg_sqs,
+                            at::TensorList state_steps,
+                            const at::Tensor& lr,
+                            const double beta1,
+                            const double beta2,
+                            const double weight_decay,
+                            const double eps,
+                            const bool maximize,
+                            const c10::optional<at::Tensor>& grad_scale,
+                            const c10::optional<at::Tensor>& found_inf) {
+  std::vector<std::vector<at::Tensor>> tensor_lists{params.vec(), grads.vec(), exp_avgs.vec(), exp_avg_sqs.vec()};
+
+  const std::string kernel_name = "fused_adamw_" + scalarToMetalTypeString(params[0].scalar_type()) + "_" +
+      scalarToMetalTypeString(state_steps[0].scalar_type());
+
+  multi_tensor_apply_for_fused_optimizer<4, 512>(kernel_name,
```

Comment:
There were a lint rule at some point about prohibition of magic constants. What 4 and 512 stands for here?

Reply:
4 stands for the length of the tensor list, it helps the multi_tensor_apply be general.

512 specifies num threads in a threadgroup.